### PR TITLE
Sentinel 2pc controller bug fixes

### LINKED ISSUE #140: Accessing nonexistent elements in sentinel_2pc::controller
### Affected Branch

trunk

### Basic Diagnostics

- [X] I've pulled the latest changes on the affected branch and the issue is still present.

- [X] The issue is reproducible in docker

### Description

In sentinel_2pc::controller,  a nonexistent element in the vector `m_opts.m_sentinel_endpoints` can be accessed if `m_sentinel_id > m_opts.m_sentinel_endpoints.size() - 1`.  The access occurs in Line 51 of src/ush/twophase/sentinel_2pc/controller.cpp:
```c++
if(ep == m_opts.m_sentinel_endpoints[m_sentinel_id]) {
```
and in Line 70:
```c++
m_opts.m_sentinel_endpoints[m_sentinel_id]);
```
Accessing a nonexistent element via the [] operator results in undefined behavior.  In my tests, the 1st access (Line 51) doesn't throw an error, but the 2nd access sometimes throws an error.  To trigger the nonexistent-element access, you can add this unit test to controller_test.cpp:
```c++
TEST_F(sentinel_2pc_test, out_of_range_sentinel_id) {
    constexpr uint32_t bad_sentinel_id = 1; 

    // Add private key for the bad sentinel ID to avoid triggering the error 
    // "No private key specified".
    constexpr auto sentinel_private_key
        = "000000000000000100000000000000000000000000000000000000000000000"
          "1";
    m_opts.m_sentinel_private_keys[bad_sentinel_id]
        = cbdc::hash_from_hex(sentinel_private_key);

    auto ctl = std::make_unique<cbdc::sentinel_2pc::controller>(bad_sentinel_id,
                                                                m_opts,
                                                                m_logger);
    ASSERT_FALSE(ctl->init());
}
```

### Solution
A solution is to check whether `sentinel_id` is out of range of `m_opts.m_sentinel_endpoints` in sentinel_2pc::controller::init, and if it is, indicate that initialization failed by returning false.  Something like this should work:
```c++
if(m_sentinel_id > (m_opts.m_sentinel_endpoints.size()-1)) {
    m_logger->error(“Sentinel ID is out of range of the sentinel endpoints vector”);
    return false;
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #141: Non-existent element access and infinite loop in sentinel_2pc::controller
### Affected Branch

trunk

### Basic Diagnostics

- [X] I've pulled the latest changes on the affected branch and the issue is still present.

- [X] The issue is reproducible in docker

### Description

# Problems
I’ve observed 2 issues with `sentinel_2pc::controller::gather_attestations`:  a non-existent element access and an infinite loop.

### 1. Non-existent element access
When `m_sentinel_endpoints.size() = 1` and `m_attestation_threshold > 1`, a non-existent element in the vector `m_sentinel_clients` is accessed.   The access results in undefined behavior, but in my tests, a seg. fault is usually thrown.  The access occurs on Line 166 in src/uhs/twophase/sentinel_2pc/controller.cpp (lines 165 - 176 shown below):
```c++
                success
                    = m_sentinel_clients[sentinel_id]->validate_transaction(
                        tx,
                        [=](validate_result v_res) {
                            auto r = requested;
                            r.insert(sentinel_id);
                            validate_result_handler(v_res,
                                                    tx,
                                                    result_callback,
                                                    ctx,
                                                    r);
                        });
```
To reproduce the error, add this test to controller_test.cpp:
```c++
TEST_F(sentinel_2pc_test, nonexistent_element_access_in_gather_attestations) {
    m_opts.m_attestation_threshold = 2;
    auto ctl = std::make_unique<cbdc::sentinel_2pc::controller>(0,
                                                                m_opts,
                                                                m_logger);
    ASSERT_TRUE(ctl->init());

    // The next line causes a non-existent element to be accessed when 
    // execute_transaction calls gather_attestations.
    auto res = ctl->execute_transaction(m_valid_tx, [](auto /* param */) {});
    ASSERT_TRUE(res);
}
```

## 2.  Infinite loop   

When the number of required sentinel attestations is larger than the number of sentinels, an infinite loop can be triggered in `sentinel_2pc::controller:;gather_attestations`. The infinite loop occurs because the condition on Line 158 in src/uhs/twophase/sentinel_2pc/controller.cpp can never be satisfied:
```
 if(ctx.m_attestations.size() < m_opts.m_attestation_threshold) {
```
Sometimes the loop will eventually throw a seg. fault, but I don’t know why yet.   

To reproduce the problem, add this test to controller_test.cpp:
```c++
TEST_F(sentinel_2pc_test, execute_transaction_triggers_infinite_loop) {
    // Require 3 sentinel attestations.  This requirement is impossible to
    // achieve since this test only launches 2 sentinels.
    m_opts.m_attestation_threshold = 3;

    // Define 2nd sentinel controller (the 1st is defined in the fixture).
    unsigned short sentinel_port = 32003;
    auto sentinel_ep
        = std::make_pair(cbdc::network::localhost, sentinel_port);
    m_opts.m_sentinel_endpoints.push_back(sentinel_ep);
    constexpr auto sentinel_private_key
        = "000000000000000100000000000000000000000000000000000000000000000"
            "1";
    constexpr auto sentinel_public_key
        = "eaa649f21f51bdbae7be4ae34ce6e5217a58fdce7f47f9aa7f3b58fa2120e2b"
            "4";
    m_opts.m_sentinel_private_keys[1]
        = cbdc::hash_from_hex(sentinel_private_key);
    m_opts.m_sentinel_public_keys.insert(
        cbdc::hash_from_hex(sentinel_public_key));

    // Initialize the 1st sentinel controller.
    auto ctl = std::make_unique<cbdc::sentinel_2pc::controller>(0,
                                                                m_opts,
                                                                m_logger);
    ASSERT_TRUE(ctl->init());


    // Initialize the 2nd sentinel controller.
    auto ctl2 = std::make_unique<cbdc::sentinel_2pc::controller>(1,
                                                            m_opts,
                                                            m_logger);
    ASSERT_TRUE(ctl2->init());

    // This line will trigger an infinite loop when execute transaction calls
    // gather_attestations.
    auto res = ctl->execute_transaction(m_valid_tx, [](auto /* param */) {});
    ASSERT_TRUE(res);
}

```

# Solutions
To prevent these errors, we could add two bits of code to `sentinel_2pc::controller::init`.

1.  We could prevent accidental casting that occurs when defining the distribution `m_dist` (Line 57 in src/uhs/twophase/sentinel_2pc/controller.cpp):
```c++
m_dist = decltype(m_dist)(0, m_sentinel_clients.size() - 1);
```
When `m_sentinel_clients.size() = 0`, the second argument to `m_dist` above is -1, which is then cast to an unsigned integer.  In my tests, it  becomes 18446744073709551615.  To prevent this, we could replace the current definition of `m_dist` with:
```c++
constexpr size_t dist_lower_bound = 0;
const size_t dist_upper_bound 
	= m_sentinel_clients.empty() ? 0 : m_sentinel_clients.size() - 1;
m_dist = decltype(m_dist)(dist_lower_bound, dist_upper_bound);
```
This ensures that the `sentinel_id` defined on Line 161 in `sentinel_2pc::controller::gather_attestations` remains a reasonable number:
```c++
auto sentinel_id = m_dist(m_rand);
```

2.  We could add the following to `sentinel_2pc::controller::init` to check that the required number of attestations never exceeds the number of sentinels:
```c++
        if(m_opts.m_attestation_threshold
            > m_opts.m_sentinel_endpoints.size()) {
            m_logger->error("The number of required attestations is larger \n \
                than the number of sentinels that can provide them.");
            return false;
        } 
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mszulcz-mitre**
This PR addresses the bugs identified in Issue #140 and Issue #141.  The bug fixes and associated unit tests are the same as those described in the issues.

### CODE REVIEW

File: src/uhs/twophase/sentinel_2pc/controller.cpp [hunk 1]
```diff
@@ -25,6 +25,20 @@ namespace cbdc::sentinel_2pc {
                                                    .size())]) {}
 
     auto controller::init() -> bool {
+        if(m_sentinel_id > (m_opts.m_sentinel_endpoints.size() - 1)) {
+            m_logger->error(
+                "The sentinel ID is too large for the number of sentinels.");
+            return false;
+        }
+
+        if(m_opts.m_attestation_threshold
```
Comment (metalicjames) [inline src/uhs/twophase/sentinel_2pc/controller.cpp]: I think this check more appropriately belongs in the configuration file parsing code where there are other sanity checks like this.

Comment (mszulcz-mitre) [reply src/uhs/twophase/sentinel_2pc/controller.cpp]: Good call.  In the latest commit, I put it with the other sanity checks on the `options` struct in the function `cbdc::config::check_options`.

Comment (HalosGhost) [reply src/uhs/twophase/sentinel_2pc/controller.cpp]: For my own clarity, which check got moved?

Comment (HalosGhost) [reply src/uhs/twophase/sentinel_2pc/controller.cpp]: Pardon the noise. The way github is displaying this is quite strange.

File: src/util/common/config.cpp [hunk 2]
```diff
@@ -682,6 +682,11 @@ namespace cbdc::config {
                 return "Two-phase mode requires at least one configured "
                        "sentinel";
             }
+            if(opts.m_sentinel_endpoints.size()
```
Comment (HalosGhost) [inline src/util/common/config.cpp:688]: It looks like this commit added this check but didn't remove it from `::init()`; is it valuable having this check twice?

Comment (mszulcz-mitre) [reply src/util/common/config.cpp:688]: Nope--it's probably not valuable.  I just forgot about it.  I just refactored the commits to remove it and to hopefully improve the commit messages.

**Comment by HalosGhost** [comment]
@mszulcz-mitre this has a very minimal merge conflict with `trunk` (just the addition of some tests). When you have a moment and can rebase it, I'll give it one final review and we'll get this merged as well!

**Comment by HalosGhost** [comment]
@metalicjames when you have a chance to review this, I'd love a second set of eyes.

File: src/uhs/twophase/sentinel_2pc/controller.cpp [hunk 3]
```diff
@@ -25,6 +25,12 @@ namespace cbdc::sentinel_2pc {
                                                    .size())]) {}
 
     auto controller::init() -> bool {
+        if(m_sentinel_id > (m_opts.m_sentinel_endpoints.size() - 1)) {
```
Comment (metalicjames) [inline src/uhs/twophase/sentinel_2pc/controller.cpp]: Does this check still work if `m_sentinel_endpoints.size()` is zero? Does `m_opts.m_sentinel_endpoints.size() - 1` result in underflow, bypassing the check?

**Review by metalicjames — COMMENTED** [review]
I think we should add a unit test for the potential underflow case, then I think it's good to merge.

**Comment by mszulcz-mitre** [comment]
I modified the code to avoid the possibility of underflow.  For clarity, I added another check to confirm that at least one sentinel endpoint is defined and I added a related unit test too.

**Review by metalicjames — APPROVED** [review]

**Comment by mszulcz-mitre** [comment]
@HalosGhost It looks like James approved the changes I made in response to his review.  Would you like me to do anything more for this pull request?

**Comment by HalosGhost** [comment]
Sorry for the delay; I've finally had a chance to walk through the code locally and test it. Tested ACK. All these changes eliminate several failure points and do so minimally. Looks good to me.

**Pull request merged**